src/cc/arduino/plugins/wifi101/flashers/java/NinaFlasher.java

```diff
@@ -0,0 +1,241 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As a special exception, you may use this file as part of a free software
+ * library without restriction.  Specifically, if other files instantiate
+ * templates or use macros or inline functions from this file, or you compile
+ * this file and link it with other files to produce an executable, this
+ * file does not by itself cause the resulting executable to be covered by
+ * the GNU General Public License.  This exception does not however
+ * invalidate any other reasons why the executable file might be covered by
+ * the GNU General Public License.
+ */
+package cc.arduino.plugins.wifi101.flashers.java;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.*;
+import java.util.List;
+import java.io.ByteArrayOutputStream;
+import java.io.*;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import org.apache.commons.lang3.StringUtils;
+import java.security.MessageDigest;
+
+import cc.arduino.plugins.wifi101.certs.WiFi101Certificate;
+import cc.arduino.plugins.wifi101.certs.WiFi101CertificateBundle;
+
+import cc.arduino.plugins.wifi101.flashers.Flasher;
+import javax.swing.JProgressBar;
+
+public class NinaFlasher implements Flasher {
+
+	public String modulename;
+	public String version;
+	public File file;
+	public JProgressBar progressBar;
+	public String name;
+	public ArrayList<String> compatibleBoard;
+	public byte[] md5Checksum;
+
+	public NinaFlasher(String _modulename, String _version, String _filename, ArrayList<String> _compatibleBoard) {
+		modulename = _modulename;
+		compatibleBoard = new ArrayList<String>();
+		version = _version;
+		file = null;
+		name = "NINA";
+
+		try {
+			String jarPath = Flasher.class.getProtectionDomain().getCodeSource().getLocation().toURI().getPath();
+			File jarFolder = new File(jarPath).getParentFile();
+			file = new File(jarFolder, _filename);
+			md5Checksum = getMD5Checksum(file.getPath());
+			compatibleBoard.addAll(_compatibleBoard);
+		}
+		catch (URISyntaxException e) {
+			e.printStackTrace();
+		}
+		catch (Exception e) {
+			e.printStackTrace();
+		}
+	}
+
+	@Override
+	public void testConnection(String port) throws Exception {
+		FlasherSerialClient client = null;
+		try {
+			progress(50, "Testing programmer...");
+			client = new FlasherSerialClient();
+			client.open(port);
+			client.hello();
+			progress(100, "Done!");
+		} finally {
+			try {
+				if (client != null)
+					client.close();
+			} catch (Exception e) {
+				e.printStackTrace();
```

Comment:
Is this how errors are handled before?

Reply:
yes it is: https://github.com/arduino-libraries/WiFi101-FirmwareUpdater-Plugin/blob/master/src/cc/arduino/plugins/wifi101/flashers/java/JavaFlasher.java#L57